### PR TITLE
Add a basic performance visualization a `/$perf`

### PR DESCRIPTION
**Pull request by jakemac53**
Fixes https://github.com/dart-lang/build/issues/915

For now incremental builds are a lot more interesting to look at, because they do a better job of capturing time spent in any individual actual builder (previous inputs will be built as a part of the `setup` phase).

cc @matanlurey @alorenzen 

[examples.zip](https://github.com/dart-lang/build/files/1712132/examples.zip)

### CODE REVIEW

**Comment by jakemac53** [comment]
You can also just save the web page and attach that to github issues, so I did that with a full and incremental build for angular_components_example here to illustrate how its done (and so you can easily see what it looks like).

File: build_runner/lib/src/generate/build_impl.dart [hunk 1]
```diff
@@ -122,6 +122,8 @@ class BuildImpl {
   final bool _verbose;
   final BuildEnvironment _environment;
 
+  BuildPerformanceTracker _performanceTracker;
```
Comment (natebosch) [inline build_runner/lib/src/generate/build_impl.dart]: non-final fields in this class make me a little nervous... I'm not sure I have a better suggestion though and at least its private 🤷‍♂️

**Comment by jakemac53** [comment]
Oh I forgot to add even the most basic of tests :D, ill go ahead and at least add some unit tests really quick.

File: build_runner/lib/src/generate/performance_tracker.dart [hunk 2]
```diff
@@ -19,45 +19,82 @@ abstract class Timings {
 
 /// The [Timings] of an entire build, including all its [actions].
 abstract class BuildPerformance implements Timings {
-  /// The [Timings] of each [BuildAction] ran in this build.
-  Iterable<BuildActionPerformance> get actions;
+  /// The [Timings] of each phase ran in this build.
+  Iterable<BuildPhasePerformance> get phases;
```
Comment (natebosch) [inline build_runner/lib/src/generate/performance_tracker.dart]: were you planning on adding an `@experimental` annotation to this library?

File: build_runner/lib/src/server/server.dart [hunk 3]
```diff
@@ -120,3 +142,85 @@ class AssetHandler {
...
+    for (var phase in action.phases) {
+      var start = phase.startTime.millisecondsSinceEpoch -
+          performance.startTime.millisecondsSinceEpoch;
+      var end = phase.stopTime.millisecondsSinceEpoch -
+          performance.startTime.millisecondsSinceEpoch;
+
+      String phaseName;
+      switch (phase.phase) {
+        case BuilderActionPhase.Setup:
+          phaseName = "setup";
```
Comment (natebosch) [inline build_runner/lib/src/server/server.dart]: inconsistent quotes in this file

File: build_runner/lib/src/server/server.dart [hunk 4]
```diff
@@ -120,3 +142,85 @@ class AssetHandler {
...
+        dataTable.addColumn({ type: 'string', id: 'Phase' });
+        dataTable.addColumn({ type: 'number', id: 'Start' });
+        dataTable.addColumn({ type: 'number', id: 'End' });
+        dataTable.addRows([
+$rows
+        ]);
+
+        var options = {
+          colors: ['#cbb69d', '#603913', '#c69c6e']
+          // timeline: {
```
Comment (natebosch) [inline build_runner/lib/src/server/server.dart]: delete

File: build_runner/lib/src/server/server.dart (hunk 3 above)
Comment (jakemac53) [reply build_runner/lib/src/server/server.dart]: done

File: build_runner/lib/src/generate/build_impl.dart (hunk 1 above)
Comment (jakemac53) [reply build_runner/lib/src/generate/build_impl.dart]: Ya, it isn't ideal. We actually have a lot of these that have to get cleaned up after each build (_lazyPhases, _resourceManager).

The real solution is probably to create another class that handles running a single build, but we can do that separately.

File: build_runner/lib/src/generate/performance_tracker.dart (hunk 2 above)
Comment (jakemac53) [reply build_runner/lib/src/generate/performance_tracker.dart]: done

**Pull request merged**